src/tfr_dumping/src/dumping_action_server.cpp

```diff
@@ -0,0 +1,164 @@
+#include <dumping_action_server.h>
+
+/*
+ * The dumping action server, it backs up the rover into the navigational aid
+ * slowly.
+ *
+ * Its first step is to make sure it can see the aruco board, it will abort the
+ * mission if it can't. 
+ *
+ * I backs up at a set speed until it get's really close and loses sight of the
+ * board. When it is blind, it drives straight back, goes slower. 
+ *
+ * It stops when the light detector get's triggered.
+ *
+ * It requires a service where it can get the most recent image on demand for
+ * the camera of interest for backing up. 
+ * */
+Dumper::Dumper(ros::NodeHandle &node, const std::string &service_name,
+        const DumpingConstraints &c) :
+    server{node, "dump", boost::bind(&Dumper::dump, this, _1), false},
+    detector{"light_detection"},
+    //TODO add dumping controller here
+    //TODO add aruco here
+    constraints{c}
+{
+    ROS_INFO("dumping action server initializing");
+    detector.waitForServer();
+
+    image_client = node.serviceClient<tfr_msgs::WrappedImage>(service_name);
+
+    velocity_publisher = node.advertise<geometry_msgs::Twist>("cmd_vel", 10);
+
+    server.start();
+    ROS_INFO("dumping action server initialized");
+}
+
+/*
+ * The business logic of the action server.
+ */
+void Dumper::dump(const tfr_msgs::EmptyGoalConstPtr &goal) 
+{  
+    //check to make sure we can see the board
+    auto initial_estimate = get_aruco_estimate();
+    if (initial_estimate.markers_detected == 0)
+    {
+        server.setAborted();
+        return;
+    }
+
+    //initialize
+    tfr_msgs::EmptyGoal empty_goal{};
+    detector.sendGoal(empty_goal);
+
+    //loop until we see the light
+    while (detector.getState() != actionlib::SimpleClientGoalState::SUCCEEDED)
+    {
+        //handle preemption
+        if (server.isPreemptRequested() || !ros::ok())
+        {
+            stop_moving();
+            server.setPreempted();
+            return;
+        }
+
+        //get the most recent aruco reading
+        tfr_msgs::ArucoResult estimate = get_aruco_estimate();
+
+        //send motor commands
+        if (estimate.markers_detected == 0)
+            move_blind();
+        else
+        {
+            update_control_msg(estimate);
+            velocity_publisher.publish(cmd);
+        }
+    }
+    stop_moving();
+    server.setSucceeded();
```

Comment:
Presumably this is where we'd add the call to the BinController to actually dump?  If so, let's add a `TODO` comment.

Reply:
Sure will do!